flow/ThreadHelper.actor.h

```diff
@@ -69,50 +71,110 @@ void onMainThreadVoid(F f, Error* err = nullptr, TaskPriority taskID = TaskPrior
 	g_network->onMainThread(std::move(signal), taskID);
 }
 
+class ThreadMultiCallback;
+
 struct ThreadCallback {
 	virtual bool canFire(int notMadeActive) const = 0;
 	virtual void fire(const Void& unused, int& userParam) = 0;
 	virtual void error(const Error&, int& userParam) = 0;
 	virtual ThreadCallback* addCallback(ThreadCallback* cb);
 
-	virtual bool contains(ThreadCallback* cb) const { return false; }
-
 	virtual void clearCallback(ThreadCallback* cb) {
 		// If this is the only registered callback this will be called with (possibly) arbitrary pointers
 	}
 
 	virtual void destroy() { UNSTOPPABLE_ASSERT(false); }
 	virtual bool isMultiCallback() const { return false; }
+
+	// MultiCallbackHolder is a helper object for ThreadMultiCallback which allows it to store its index
+	// within the callback vector inside the ThreadCallback rather than having a map of pointers or
+	// some other scheme to store the indices by callback.
+	// MultiCallbackHolder objects can form a doubly linked list.
+	struct MultiCallbackHolder : public FastAllocated<MultiCallbackHolder> {
+		MultiCallbackHolder(ThreadMultiCallback* holder = nullptr,
+		                    MultiCallbackHolder* prev = nullptr,
+		                    MultiCallbackHolder* next = nullptr)
+		  : holder(holder), previous(prev), next(next) {}
+
+		ThreadMultiCallback* holder;
+		int index;
+		MultiCallbackHolder* previous;
+		MultiCallbackHolder* next;
+	};
+
+	// firstHolder is both the inline first record of a MultiCallbackHolder and the head of the
+	// doubly linked list of MultiCallbackHolder entries.
+	MultiCallbackHolder firstHolder;
+
+	// Return a MultiCallbackHolder for the given holder, using the firstHolder if free or allocating
+	// a new one.  No check for an existing record for holder is done.
+	MultiCallbackHolder* addHolder(ThreadMultiCallback* holder) {
+		if (firstHolder.holder == nullptr) {
+			firstHolder.holder = holder;
+			return &firstHolder;
+		}
+		firstHolder.next = new MultiCallbackHolder(holder, &firstHolder, firstHolder.next);
+		return firstHolder.next;
+	}
+
+	// Get the MultiCallbackHolder for holder if it exists, or nullptr.
+	MultiCallbackHolder* getHolder(ThreadMultiCallback* holder) {
```

Comment:
how many callbacks does it usually have? wondering if we shall calculate the holder address from the ThreadMultiCallback address.

Reply:
The linked list will rarely have more than 2 entries, and often just 1 which is why the first linked list entry is inline in ThreadCallback.  This is actually from a different PR #6721 which was needed here for determinism debugging, see the other PR for details.